src/wayland/shell/legacy/decoration.rs

```diff
@@ -0,0 +1,128 @@
+//!
+//! ```no_run
+//! # extern crate wayland_server;
+//! #
+//! use smithay::wayland::shell::legacy::decoration::{init_kde_decoration_manager, KdeDecorationRequest};
+//! use wayland_protocols::misc::server_decoration::server::org_kde_kwin_server_decoration_manager::Mode;
+//!
+//! # let mut display = wayland_server::Display::new();
+//!
+//! let decoration_manager = init_kde_decoration_manager(
+//!     &mut display,
+//!     |decoration, request| match request {
+//!         KdeDecorationRequest::Setup { manager } => manager.default_mode(Mode::Server),
+//!         KdeDecorationRequest::NewToplevelDecoration { surface } => (),
+//!         KdeDecorationRequest::RequestMode { mode, .. } => decoration.mode(mode),
+//!         KdeDecorationRequest::Release => (),
+//!     },
+//!     None,
+//! );
+
+use std::{cell::RefCell, ops::Deref, rc::Rc};
+
+use wayland_protocols::misc::server_decoration::server::{
+    org_kde_kwin_server_decoration::{Mode, OrgKdeKwinServerDecoration, Request as ServerDecorationRequest},
+    org_kde_kwin_server_decoration_manager::{
+        OrgKdeKwinServerDecorationManager, Request as ServerDecorationManagerRequest,
+    },
+};
+use wayland_server::protocol::wl_surface::WlSurface;
+use wayland_server::{Display, Filter, Global, Main};
+
+/// Events generated by the kde decoration manager.
+#[derive(Debug)]
+pub enum KdeDecorationRequest {
+    /// Decoration manager setup completion.
+    Setup {
+        /// KDE decoration manager.
+        manager: OrgKdeKwinServerDecorationManager,
+    },
+    /// A new toplevel decoration was instantiated.
+    NewToplevelDecoration {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+    /// Informs the compositor that the client prefers the provided decoration mode.
+    RequestMode {
+        /// KDE decoration object.
+        decoration: OrgKdeKwinServerDecoration,
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+        /// Desired mode.
+        mode: Mode,
+    },
+    /// Decoration object was released.
+    Release {
+        /// The surface asosiated with the decoration.
+        surface: WlSurface,
+    },
+}
+
+/// Create a new KDE Decoration Manager global
+///
+/// A compositor can set the default mode for new clients at any time using
+/// [`OrgKdeKwinServerDecorationManager::default_mode`]. For individual windows the decorations can
+/// be set using [`OrgKdeKwinServerDecoration::mode`]. Both of these requests will cause the client
+/// to respond with its desired mode using [`KdeDecorationRequest::RequestMode`], which then should
+/// be acknowledged by the compositor using another [`OrgKdeKwinServerDecoration::mode`] event.
+/// **It is up to the compositor to prevent feedback loops**, a client is free to ignore the
+/// suggested decoration mode.
+pub fn init_kde_decoration_manager<L, Impl>(
```

Comment:
I think it would make more sense to have init function take the default mode as a parameter. The return value can be adjusted to also return some object that may be used to change this default mode at any time.